### PR TITLE
Support JSON values for secretKey

### LINKED ISSUE #115: fetching complex JSON from Vault
from the changelog...

```
If secretKey is set, that is the key from the secret's data that will be written
If secretKey is not set, the whole JSON response from Vault will be written
```

I have complex JSON data stored in Vault, example (heavily redacted, use your imagination):

image

the data shows up correctly when `secretKey` is excluded. however, the "whole JSON response" shows up and it includes extra details, ex:

```
$ head /mnt/secrets-store/fabricProfile 
{"request_id":"deadbeef-xxxx-yyyy-zzzz-moreredactedhex","lease_id":"","lease_duration":0,"renewable":false,"data":{"data":{"certificateAuthorities":{"hlf-ca--unicorn":{....
```

I tried setting `secretKey` to a variety of things like `data` `Data` `.data` to see if I could trick the provider into sending me the JSON body without the extraneous response "envelope".

is there a specific key I can use when fetching JSON blocks? I was really hoping to leverage JSON instead of only key/value pairs. or is this something that would require code-changes to the `vault-csi-provider` to support?

### PR DESCRIPTION
**Pull request by tomhjp**
Resolves #115 and based on code from #122. You can put arbitrary JSON data into `kv`'s values (and theoretically other secret engines could do the same) using a JSON file and a command like `vault kv put secret/test @test.json`, and this PR allows `vault-csi-provider` to specifically select that JSON data without having to escape your secret JSON into a big ugly string when storing it in Vault.

This PR does _not_ address arbitrary paths within the JSON though. Vault's API itself enforces that the top-level data object is a map, and users should be able to select any single top-level key from that object.

Strings remain special-cased so that we continue to write string values without wrapping them in quotes.

### CODE REVIEW

**Comment by agates4** [comment]
Hey Tom, this looks great to me. Thanks for the thorough changes and response!

File: internal/provider/provider.go [hunk 1]
```diff
@@ -164,35 +164,48 @@ func keyFromData(rootData map[string]interface{}, secretKey string) (string, err
...
+		return []byte(content), nil
+	}
+
+	// Arbitrary data can be returned in the data field of an API secret struct.
+	// It's already been Unmarshalled from the response, so in theory,
+	// marshalling should never realistically fail, but don't log the error just
+	// in case, as it could contain secret contents if it does somehow fail.
+	if content, err := json.Marshal(data[secretKey]); err == nil {
+		return content, nil
 	}
```
Comment (tvoran) [inline internal/provider/provider.go:179]: Interesting, I naively thought these would be flipped (checking for json and then string), but of course data is already unmarshalled here.

File: internal/provider/provider.go [hunk 2]
```diff
@@ -164,35 +164,48 @@ func keyFromData(rootData map[string]interface{}, secretKey string) (string, err
...
-			return "", fmt.Errorf("couldn't read secret %q: %w", secretConfig.ObjectName, err)
+			return nil, fmt.Errorf("couldn't read secret %q: %w", secretConfig.ObjectName, err)
 		}
 		if secret == nil || secret.Data == nil {
-			return "", fmt.Errorf("empty response from %q, warnings: %v", req.URL.Path, secret.Warnings)
+			return nil, fmt.Errorf("empty response from %q, warnings: %v", req.URL.Path, secret.Warnings)
+		}
+
+		for _, w := range secret.Warnings {
+			p.logger.Warn("warning in response from Vault API", "warning", w)
```
Comment (tvoran) [inline internal/provider/provider.go:208]: Nice addition!

**Review by tvoran — APPROVED** [review]

File: internal/provider/provider.go (hunk 1 above)
Comment (tomhjp) [reply internal/provider/provider.go:179]: Yeah it took me a surprising amount of iteration to arrive at this. I wanted to just have JSON, but then the extra quotes ruin the most common usage.

**Pull request merged**